### PR TITLE
Implement some more slice allocation functions

### PR DESCRIPTION
**Pull request by TethysSvensson**
This implements the feature requested in #41.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -595,6 +615,114 @@ impl Bump {
...
+    /// ## Example
+    ///
+    /// ```
+    /// let bump = bumpalo::Bump::new();
+    /// let x = bump.alloc_slice_fill_copy(5, 42);
+    /// assert_eq!(x, &[42, 42, 42, 42, 42]);
+    /// ```
+    #[inline(always)]
+    #[allow(clippy::mut_from_ref)]
+    pub fn alloc_slice_fill_copy<T: Copy>(&self, len: usize, value: T) -> &mut [T] {
```
Comment (ztlpn) [inline src/lib.rs]: Hey @TethysSvensson thanks for implementing the feature I requested, that was quick! One remark: I messed up the order of arguments in the `vec![]` macro in my request (as I always do, it seems impossible to remember!) - the correct order is `vec![value, len]`. So I think here we should use the same order fo consistency.

Comment (TethysSvensson) [reply src/lib.rs]: You are right, that makes more sense.

**Comment by TethysSvensson** [comment]
> Implementation looks great -- can you also add some tests for these new methods? Thanks!

Do you mean beyond the ones in the doctests? Which tests would you like me to implement?

**Comment by fitzgen** [comment]
> Do you mean beyond the ones in the doctests? Which tests would you like me to implement?

* tests for the `len == 0` fast path

* tests for when the layout overflows in rounding up and in the multiply

* would be good to get some quickchecks in `tests/quickchecks.rs` as well:

```rust
quickcheck! {
    // ...

    fn alloc_slices(allocs: Vec<(u8, usize)>) -> () {
        let b = Bump::new();
        let mut allocated = vec![];
        for (val, len) in allocs {
            let s = b.alloc_slice_copy(val, len); // maybe need to place a max on `len`.

            // todo: assert that the slice is all `val`

            let range = (s.as_ptr(), s.as_ptr().add(s.len()));
            for r in &allocated {
                assert!(!overlaps(*r, range));
            }
            allocated.push(range);
        }
    }
}
```

Thanks!

**Comment by TethysSvensson** [comment]
I think we should also provide a function that takes a closure and calls it for every element. That would also be useful internally to avoid code duplication.

**Comment by fitzgen** [comment]
> I think we should also provide a function that takes a closure and calls it for every element. That would also be useful internally to avoid code duplication.

:+1:

**Comment by TethysSvensson** [comment]
I have rebased this on top of master and added a new commit.

**Comment by TethysSvensson** [comment]
I am still missing more tests before this can merge.

**Comment by TethysSvensson** [comment]
@fitzgen I've implemented the tests now. Please take a look.

I also moved the special-casing of ZSTs into the core allocation instead of dealing with it in `alloc_slice`. There is a test that the allocation path works as expected. I am not sure how to test the deallocation/reallocation path. If you have any ideas, I'll be happy to add tests, but otherwise I think the code is simple enough that it works as is.

**Pull request merged**